File: Apps4Bro.Android/src/net/runserver/apps4bro/Apps4BroSDK.java
package net.runserver.apps4bro;

import android.app.Activity;
import android.app.Application;
import android.content.Context;
import android.os.Build;
import android.util.Log;

import com.google.android.gms.ads.MobileAds;
import com.google.android.gms.ads.initialization.InitializationStatus;
import com.google.android.gms.ads.initialization.OnInitializationCompleteListener;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.URLEncoder;
import java.util.Locale;

import com.google.android.gms.ads.identifier.AdvertisingIdClient;
import com.google.android.ump.ConsentForm;
import com.google.android.ump.FormError;

import androidx.annotation.Nullable;

public final class Apps4BroSDK
{
    public static final int Version = 152;

    private static String s_advertisingId = "";
    /// Cached at init() — sent as `app=` analytics name in AdManagerUrl.
    private static String s_packageName = "";
    private static boolean s_inited = false;

    private static boolean s_adTrackingLimited = true;

    private static String s_platform = "unknown";

    public final static String App4BroTag = "app4bro";
    public final static String ReportUrl = "https://app4bro.runserver.net/app4bro/event.php?id=%1$s&app=%2$s&event=%3$s&param=%4$s&time=%5$d&eventid=%6$s";
    // /route/<App4Bro-ID>/ — path-embedded routing. Query params:
    //   id    = App4Bro app ID (same value as the path segment). Routing key.
    //   app   = analytics name (package name). Cosmetic.
    //   lang/sdk/os = metadata.
    //   did   = device advertising ID. Trailing; may be empty when SDK not inited.
    public final static String AdManagerUrl = "https://app4bro.runserver.net/route/%1$s/?id=%1$s&app=%2$s&lang=%3$s&sdk=%4$s&os=%5$s&did=%6$s";
    public final static String AdManagerUrlShort = "https://app4bro.runserver.net/route/%1$s/";

    // HouseAdUrl args (in positional order):
    //   1  id         — house zone ID (NOT the App4Bro app ID)
    //   2  app        — analytics name (package name — same convention as AdManagerUrl)
    //   3  brand      — device brand (Build.MANUFACTURER)
    //   4  model      — device model (Build.MODEL)
    //   5  operator   — carrier
    //   6  width, 7 height
    //   8  lang       — ISO 639-1 language code
    //   9  sdk        — Apps4BroSDK.Version
    //   10 os         — platform name
    //   11 osver      — OS version (Build.VERSION.RELEASE)
    //   12 appver     — host app versionName (from PackageInfo)
    //   13 did        — device advertising ID (trailing; may be empty)
    public final static String HouseAdUrl = "https://app4bro.runserver.net/app4bro/house.php?id=%1$s&app=%2$s&brand=%3$s&model=%4$s&operator=%5$s&width=%6$d&height=%7$d&lang=%8$s&sdk=%9$s&os=%10$s&osver=%11$s&appver=%12$s&did=%13$s";
    public final static int HouseAdTimeout = 10;

    public static String getAdvertisingId() throws Exception
    {
        if (!s_inited)
            throw new Exception("Apps4Bro SDK is not inited!");
        return s_advertisingId;
    }

    public static String getPlatform()
    {
        return s_platform;
    }

    public static boolean isAdTrackingLimited()
    {
        return s_adTrackingLimited;
    }

    public static void init(final Context context)
    {
        s_inited = true;
        s_platform = Build.MANUFACTURER.equals("Amazon") ? "amazon" : "android";
        s_packageName = context != null ? context.getPackageName() : "";

        if (isSystemLanguageUkrainian(context))
            return;

        if (context instanceof CardApplication)
        {
            ConsentManager.getInstance(context).showPrivacyOptionsForm(((CardApplication)context).getCurrentActivity(), new ConsentForm.OnConsentFormDismissedListener()
            {
                @Override
                public void onConsentFormDismissed(@Nullable FormError formError)
                {

                }
            });
        }

        // Initialize Google advertising id
        new Thread(() ->
        {
            try
            {
                AdvertisingIdClient.Info adInfo = AdvertisingIdClient.getAdvertisingIdInfo(context);
                s_adTrackingLimited = adInfo.isLimitAdTrackingEnabled();
                if (!s_adTrackingLimited)
                    s_advertisingId = URLEncoder.encode(adInfo.getId());
            }
            catch (Exception ex)
            {
                ex.printStackTrace();
            }
        }).start();

        // Initialize Google ads app-wide
        try
        {
            MobileAds.initialize(
                    context,
                    initializationStatus -> {
                    });
        }
        catch (Exception ex)
        {
            ex.printStackTrace();
        }
    }

    private static boolean isSystemLanguageUkrainian(final Context context) {
        Locale currentLocale;
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.N) {
            currentLocale = context.getResources().getConfiguration().getLocales().get(0);
        } else {
            currentLocale = context.getResources().getConfiguration().locale;
        }
        return currentLocale.getLanguage().equals(new Locale("uk").getLanguage());
    }

//    @NonNull
    static String loadStreamText(InputStream stream)
    {
        try
        {
            BufferedReader reader = new BufferedReader(new InputStreamReader(stream, "UTF-8"), 0x8000);

            StringBuilder result = new StringBuilder();
            String line = null;

            while ((line = reader.readLine()) != null)
            {
                result.append(line);
                // Log.d("Test", "Line: " + line);
            }

            return result.toString();
        }
        catch (Exception ex)
        {
            ex.printStackTrace();

        }
        return "";
    }

    public static String getRequestUrl(String appId)
    {
        // New args (App4Bro ID in path + id=; analytics name in app=; did= trails):
        //   %1$s = App4Bro app ID
        //   %2$s = package name (analytics)
        //   %3$s = language code
        //   %4$s = SDK version
        //   %5$s = platform name
        //   %6$s = device advertising ID (may be empty)
        try
        {
            return String.format(Apps4BroSDK.AdManagerUrl,
                appId,
                s_packageName,
                Locale.getDefault().getLanguage(),
                Apps4BroSDK.Version,
                Apps4BroSDK.getPlatform(),
                s_advertisingId);
        }
        catch (Exception e)
        {
            Log.e(App4BroTag, "error: " + e.getMessage(), e);
        }

        return String.format(AdManagerUrlShort, appId);
    }
}

File: Apps4Bro.Windows/Apps4BroSDK.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using Windows.Storage;


#if __ANDROID__
using Android.Gms.Ads.Identifier;
using Android.Gms.Common;
#endif

namespace Apps4Bro
{
    public enum Apps4BroSettings
    {
        None = 0,
        ShowBannerOnTop = 1,
        UseBannerSuperview = 2
    }

    public static class Apps4BroSDK
    {
        private static readonly string s_emptyId = "00000000-0000-0000-0000-000000000000";
        public static readonly int Version = 130;

        private static string s_advertisingId = s_emptyId;
        private static bool s_inited = false;

        private static bool s_useNonPersonalizedAds = false;

        private static string s_platform = "unknown";

        internal readonly static string ReportUrl = "https://app4bro.runserver.net/app4bro/event.php?id={0}&app={1}&event={2}&param={3}&time={4}&eventid={5}";

        // /route/<App4Bro-ID>/ — path-embedded routing. Query params:
        //   id    = App4Bro app ID (same value as the path segment). Routing key.
        //   app   = analytics name (package family name). Cosmetic.
        //   lang/sdk/os = metadata.
        //   did   = device advertising ID. Trailing; may be empty.
        internal readonly static string AdManagerUrl = "https://app4bro.runserver.net/route/{0}/?id={0}&app={1}&lang={2}&sdk={3}&os={4}&did={5}";
        internal readonly static string AdManagerUrlShort = "https://app4bro.runserver.net/route/{0}/";

        // HouseAdUrl args:
        //   {0}  id         — house zone ID (NOT the App4Bro app ID)
        //   {1}  app        — analytics name (package family name — same as AdManagerUrl)
        //   {2}  brand
        //   {3}  model
        //   {4}  operator
        //   {5}  width, {6} height
        //   {7}  lang
        //   {8}  sdk        — Apps4BroSDK.Version
        //   {9}  os
        //   {10} osver      — OS version (AnalyticsInfo.VersionInfo.DeviceFamilyVersion)
        //   {11} appver     — host app version (Package.Current.Id.Version stringified)
        //   {12} did        — device advertising ID (trailing; may be empty)
        //   {13} noclose    — banner vs interstitial close-button flag
        internal readonly static string HouseAdUrl = "https://app4bro.runserver.net/app4bro/house.php?id={0}&app={1}&brand={2}&model={3}&operator={4}&width={5}&height={6}&lang={7}&sdk={8}&os={9}&osver={10}&appver={11}&did={12}&noclose={13}";
        internal readonly static int HouseAdTimeout = 10;

        internal static IDictionary<Apps4BroSettings, object> Settings = new Dictionary<Apps4BroSettings, object>();

        public static string Platform
        {
            get { return s_platform; }
        }

        public static bool UseNonPersonalizedAds
        {
            get { return s_useNonPersonalizedAds; }
            set { s_useNonPersonalizedAds = value; }
        }

        public static string AdvertisingId
        {
            get
            {

                if (!s_inited)
                {
#if __IOS__ || NETFX_CORE
                    Init(null); // we can init iOS sdk without context. Not possible for Android
#else
                    throw new ApplicationException("Apps4Bro SDK is not inited!");
#endif
                }

                if (string.IsNullOrEmpty(s_advertisingId) || s_advertisingId.Equals(s_emptyId))
                    return string.Empty; // conserve the traffic 

                return s_advertisingId;

            }
        }

        public static void SetSettings(Apps4BroSettings setting, object value)
        {
            Settings[setting] = value;
        }

        public static void Init(object context)
        {
            s_inited = true;
            try
            {
#if __IOS__
                
                s_platform = "ios";
                if (AdSupport.ASIdentifierManager.SharedManager.IsAdvertisingTrackingEnabled)
                        s_advertisingId = WebUtility.UrlEncode(AdSupport.ASIdentifierManager.SharedManager.AdvertisingIdentifier.AsString());
                
#elif __ANDROID__

                if (context == null)
                    throw new ApplicationException("Apps4Bro SDK is not explicitly inited!");
                
                s_platform = Adroid.OS.Build.MANUFACTURER == "Amazon" ? "amazon" : "android";
                
                System.Threading.Thread thread= new System.Threading.Thread(delegate()
                    {
                        try
                        {
                            AdvertisingIdClient.Info adInfo = AdvertisingIdClient.GetAdvertisingIdInfo((Android.Content.Context)context);
                            s_advertisingId = WebUtility.UrlEncode(adInfo.Id);
                        }
                        catch(Exception ex)
                        {
                            Console.Error.WriteLine("Failed to get Advertising ID: " + ex);
                        }
                    }
                );
                thread.Start();
#elif NETFX_CORE

                object id;

                if (!ApplicationData.Current.LocalSettings.Values.TryGetValue("did", out id))
                {
                    s_advertisingId = Guid.NewGuid().ToString("N");
                    ApplicationData.Current.LocalSettings.Values["did"] = s_advertisingId;
                }
                else
                    s_advertisingId = id.ToString();

				s_platform = "win";
#endif
            }
            catch (Exception ex)
            {
                Debug.WriteLine("Failed to get Advertising ID: " + ex);
            }
        }
    }
}

